### PR TITLE
Psse Convert PSSE model to PowSyBl model

### PR DESCRIPTION
**Pull request by marqueslanauja**
**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem ?** *If so, link to this issue using `'#XXX'` and skip the rest*

**What kind of change does this PR introduce?** *(Bug fix, feature, docs update, ...)*
Feature

**What is the current behavior?** *(You can also link to an open issue here)*

**What is the new behavior (if this is a feature change)?**
The PSSE model is converted to a PowSyBl model

**Does this PR introduce a breaking change or deprecate an API?** *If yes, check the following:*
- [ ] The *Breaking Change* or *Deprecated* label has been added
- [ ] The migration guide has been updated in the github wiki *(What changes might users need to make in their application due to this PR?)*

**Other information**:

(if any of the questions/checkboxes don't apply, please delete them entirely)

### CODE REVIEW

File: psse/psse-converter/src/main/java/com/powsybl/psse/converter/GeneratorConverter.java [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    public void addControl(PsseBus psseBus) {
+        String busId = getBusId(psseGenerator.getI());
+        Generator generator = getNetwork().getGenerator(getGeneratorId(busId, psseGenerator));
+
+        // Add control only if generator has been created
+        if (generator == null) {
+            return;
+        }
+
+        double vnom = generator.getTerminal().getVoltageLevel().getNominalV();
```
Comment (annetill) [inline psse/psse-converter/src/main/java/com/powsybl/psse/converter/GeneratorConverter.java:69]: You should use the nominal voltage linked to regulating terminal.

File: psse/psse-converter/src/main/java/com/powsybl/psse/converter/TransformerConverter.java [hunk 2]
```diff
@@ -0,0 +1,795 @@
...
+            .setRegulating(regulating);
+
+        return regulating;
+    }
+
+    private static boolean forceRegulatingToOff(boolean regulatingForcedToOff, boolean regulating) {
+        return regulatingForcedToOff || regulating;
+    }
+
+    private static Terminal defineRegulatingTerminal(Network network, String id, PsseTransformerWinding winding) {
```
Comment (annetill) [inline psse/psse-converter/src/main/java/com/powsybl/psse/converter/TransformerConverter.java:765]: Remote control for phase tap changers and ratio tap changers are not supported in PSSE format?

File: psse/psse-converter/src/main/java/com/powsybl/psse/converter/SwitchedShuntCompensatorConverter.java [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+        String busId = getBusId(psseSwitchedShunt.getI());
+        String id = defineShuntId(psseSwitchedShunt, version);
+        ShuntCompensator shunt = getNetwork().getShuntCompensator(getShuntId(busId, id));
+
+        // Add control only if shunt has been created
+        if (shunt == null) {
+            return;
+        }
+
+        double vnom = shunt.getTerminal().getVoltageLevel().getNominalV();
```
Comment (annetill) [inline psse/psse-converter/src/main/java/com/powsybl/psse/converter/SwitchedShuntCompensatorConverter.java:83]: Same question everywhere: PSSE format does not support remote control?

File: psse/psse-converter/src/main/java/com/powsybl/psse/converter/GeneratorConverter.java (hunk 1 above)
Comment (marqueslanauja) [reply psse/psse-converter/src/main/java/com/powsybl/psse/converter/GeneratorConverter.java]: Yes, you are right. I fixed and i added a unitary test. Thanks.

File: psse/psse-converter/src/main/java/com/powsybl/psse/converter/SwitchedShuntCompensatorConverter.java (hunk 3 above)
Comment (marqueslanauja) [reply psse/psse-converter/src/main/java/com/powsybl/psse/converter/SwitchedShuntCompensatorConverter.java]: Yes, you are right. I fixed and i added a unitary test. Thanks.

File: psse/psse-converter/src/main/java/com/powsybl/psse/converter/TransformerConverter.java (hunk 2 above)
Comment (marqueslanauja) [reply psse/psse-converter/src/main/java/com/powsybl/psse/converter/TransformerConverter.java:764]: Yes, it is supported. I Fixed the conversion to support it. Unitary test added.

File: psse/psse-converter/src/main/java/com/powsybl/psse/converter/FixedShuntCompensatorConverter.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+        adder.newLinearModel()
+            .setGPerSection(powerToShuntAdmittance(psseFixedShunt.getGl(), voltageLevel.getNominalV()))
+            .setBPerSection(powerToShuntAdmittance(psseFixedShunt.getBl(), voltageLevel.getNominalV()))
+            .setMaximumSectionCount(1)
+            .add();
+        ShuntCompensator shunt = adder.add();
+
+        if (psseFixedShunt.getStatus() == 1) {
+            shunt.getTerminal().connect();
+        }
```
Comment (MioRtia) [inline psse/psse-converter/src/main/java/com/powsybl/psse/converter/FixedShuntCompensatorConverter.java:52]: You can replace this by putting above:
```java
adder.setBus(psseFixedShunt.getStatus() == 1 ? busId : null)
```

File: psse/psse-converter/src/main/java/com/powsybl/psse/converter/GeneratorConverter.java [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+                .setMaxQ(psseGenerator.getQt())
+            .add();
+
+        if (psseGenerator.getRt() != 0.0 || psseGenerator.getXt() != 0.0) {
+            LOGGER.warn("Implicit method where a transformer is specified with the generator is not supported ({})", generator.getId());
+        }
+
+        if (psseGenerator.getStat() == 1) {
+            generator.getTerminal().connect();
+        }
```
Comment (MioRtia) [inline psse/psse-converter/src/main/java/com/powsybl/psse/converter/GeneratorConverter.java:57]: Same remark

File: psse/psse-converter/src/main/java/com/powsybl/psse/converter/LineConverter.java [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+ * @author Luma Zamarreño <zamarrenolm at aia.es>
+ * @author José Antonio Marqués <marquesja at aia.es>
+ */
+public class LineConverter extends AbstractConverter {
+
+    public LineConverter(PsseNonTransformerBranch psseLine, ContainersMapping containerMapping, PerUnitContext perUnitContext, Network network, PsseVersion version) {
+        super(containerMapping, network);
+        this.psseLine = psseLine;
+        this.perUnitContext = perUnitContext;
+        this.version = version;
```
Comment (MioRtia) [inline psse/psse-converter/src/main/java/com/powsybl/psse/converter/LineConverter.java:32]: Same remark regarding `requireNonNull`

File: psse/psse-converter/src/main/java/com/powsybl/psse/converter/LineConverter.java [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+            .setG1(admittanceToEngineeringUnits(psseLine.getGi(), voltageLevel2.getNominalV(), perUnitContext.getSb()))
+            .setB1(admittanceToEngineeringUnits(psseLine.getB() * 0.5 + psseLine.getBi(), voltageLevel2.getNominalV(), perUnitContext.getSb()))
+            .setG2(admittanceToEngineeringUnits(psseLine.getGj(), voltageLevel2.getNominalV(), perUnitContext.getSb()))
+            .setB2(admittanceToEngineeringUnits(psseLine.getB() * 0.5 + psseLine.getBj(), voltageLevel2.getNominalV(), perUnitContext.getSb()))
+            .add();
+
+        if (psseLine.getSt() == 1) {
+            line.getTerminal1().connect();
+            line.getTerminal2().connect();
+        }
```
Comment (MioRtia) [inline psse/psse-converter/src/main/java/com/powsybl/psse/converter/LineConverter.java:62]: Same remark about `setBus1/2`

File: psse/psse-converter/src/main/java/com/powsybl/psse/converter/LoadConverter.java [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+        Load load = voltageLevel.newLoad()
+            .setId(getLoadId(busId))
+            .setConnectableBus(busId)
+            .setP0(psseLoad.getPl())
+            .setQ0(psseLoad.getQl())
+            .add();
+
+        if (psseLoad.getStatus() == 1) {
+            load.getTerminal().connect();
+        }
```
Comment (MioRtia) [inline psse/psse-converter/src/main/java/com/powsybl/psse/converter/LoadConverter.java:48]: Same remark about `setBus`

File: psse/psse-converter/src/main/java/com/powsybl/psse/converter/SlackConverter.java [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+
+/**
+ * @author Luma Zamarreño <zamarrenolm at aia.es>
+ * @author José Antonio Marqués <marquesja at aia.es>
+ */
+public class SlackConverter extends AbstractConverter {
+
+    public SlackConverter(List<PsseBus> psseBusList, ContainersMapping containerMapping, Network network) {
+        super(containerMapping, network);
+        this.psseBusList = psseBusList;
```
Comment (MioRtia) [inline psse/psse-converter/src/main/java/com/powsybl/psse/converter/SlackConverter.java:25]: Same remark about `requireNonNull`

File: psse/psse-converter/src/main/java/com/powsybl/psse/converter/SubstationConverter.java [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+
+/**
+ * @author Luma Zamarreño <zamarrenolm at aia.es>
+ * @author José Antonio Marqués <marquesja at aia.es>
+ */
+public class SubstationConverter extends AbstractConverter {
+
+    public SubstationConverter(PsseBus psseBus, ContainersMapping containerMapping, Network network) {
+        super(containerMapping, network);
+        this.psseBus = psseBus;
```
Comment (MioRtia) [inline psse/psse-converter/src/main/java/com/powsybl/psse/converter/SubstationConverter.java:22]: Same remark about `requireNonNull`

File: psse/psse-converter/src/main/java/com/powsybl/psse/converter/SwitchedShuntCompensatorConverter.java [hunk 11]
```diff
@@ -0,0 +1,266 @@
...
+                    .setB(powerToShuntAdmittance(shuntBlock.getB(), voltageLevel.getNominalV()))
+                    .endSection();
+            }
+        });
+        modelAdder.add();
+        ShuntCompensator shunt = adder.add();
+
+        if (psseSwitchedShunt.getStat() == 1) {
+            shunt.getTerminal().connect();
+        }
```
Comment (MioRtia) [inline psse/psse-converter/src/main/java/com/powsybl/psse/converter/SwitchedShuntCompensatorConverter.java:70]: Same remark about `setBus`

File: psse/psse-converter/src/main/java/com/powsybl/psse/converter/TransformerConverter.java [hunk 12]
```diff
@@ -0,0 +1,794 @@
...
+
+    public TransformerConverter(PsseTransformer psseTransformer, ContainersMapping containersMapping,
+        PerUnitContext perUnitContext, Network network, Map<Integer, PsseBus> busNumToPsseBus, double sbase,
+        PsseVersion version) {
+        super(containersMapping, network);
+        this.psseTransformer = psseTransformer;
+        this.busNumToPsseBus = busNumToPsseBus;
+        this.sbase = sbase;
+        this.perUnitContext = perUnitContext;
+        this.version = version;
```
Comment (MioRtia) [inline psse/psse-converter/src/main/java/com/powsybl/psse/converter/TransformerConverter.java:52]: Same remark about `requireNonNull`

File: psse/psse-converter/src/main/java/com/powsybl/psse/converter/TransformerConverter.java [hunk 13]
```diff
@@ -0,0 +1,794 @@
...
+            .setB(ysh.getImaginary())
+            .add();
+
+        tapChangerToIidm(tapChangerAdjustedYsh, twt);
+        defineOperationalLimits(twt, voltageLevel1.getNominalV(), voltageLevel2.getNominalV());
+
+        if (psseTransformer.getStat() == 1) {
+            twt.getTerminal1().connect();
+            twt.getTerminal2().connect();
+        }
```
Comment (MioRtia) [inline psse/psse-converter/src/main/java/com/powsybl/psse/converter/TransformerConverter.java:123]: Same remark about `setBus1/2`

File: psse/psse-converter/src/main/java/com/powsybl/psse/converter/TransformerConverter.java [hunk 14]
```diff
@@ -0,0 +1,794 @@
...
+                break;
+            default:
+                throw new PsseException("PSSE: Unexpected CW = " + cw);
+        }
+        return ratio;
+    }
+
+    private static double defineAngle(double ang) {
+        return ang;
+    }
```
Comment (MioRtia) [inline psse/psse-converter/src/main/java/com/powsybl/psse/converter/TransformerConverter.java:316]: Is this method necessary?

File: psse/psse-converter/src/main/java/com/powsybl/psse/converter/VoltageLevelConverter.java [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+/**
+ * @author Luma Zamarreño <zamarrenolm at aia.es>
+ * @author José Antonio Marqués <marquesja at aia.es>
+ */
+public class VoltageLevelConverter extends AbstractConverter {
+
+    public VoltageLevelConverter(PsseBus psseBus, ContainersMapping containerMapping, PerUnitContext perUnitContext, Network network) {
+        super(containerMapping, network);
+        this.psseBus = psseBus;
+        this.perUnitContext = perUnitContext;
```
Comment (MioRtia) [inline psse/psse-converter/src/main/java/com/powsybl/psse/converter/VoltageLevelConverter.java:26]: Same remark about `requireNonNull`

**Review by MioRtia — COMMENTED** [review]
Just a question about your new classes: is this necessary for them to be public or can they be package private? If they can, you should put them at the most restricted mode

File: psse/psse-converter/src/main/java/com/powsybl/psse/converter/AbstractConverter.java [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+
+/**
+ * @author Luma Zamarreño <zamarrenolm at aia.es>
+ * @author José Antonio Marqués <marquesja at aia.es>
+ */
+public abstract class AbstractConverter {
+
+    protected AbstractConverter(ContainersMapping containersMapping, Network network) {
+        this.containersMapping = containersMapping;
+        this.network = network;
```
[the remark this answers is not in the document]
Comment (marqueslanauja) [reply psse/psse-converter/src/main/java/com/powsybl/psse/converter/AbstractConverter.java]: Done

File: psse/psse-converter/src/main/java/com/powsybl/psse/converter/FixedShuntCompensatorConverter.java [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+
+/**
+ * @author Luma Zamarreño <zamarrenolm at aia.es>
+ * @author José Antonio Marqués <marquesja at aia.es>
+ */
+public class FixedShuntCompensatorConverter extends AbstractConverter {
+
+    public FixedShuntCompensatorConverter(PsseFixedShunt psseFixedShunt, ContainersMapping containerMapping, Network network) {
+        super(containerMapping, network);
+        this.psseFixedShunt = psseFixedShunt;
```
[the remark this answers is not in the document]
Comment (marqueslanauja) [reply psse/psse-converter/src/main/java/com/powsybl/psse/converter/FixedShuntCompensatorConverter.java]: Done

File: psse/psse-converter/src/main/java/com/powsybl/psse/converter/FixedShuntCompensatorConverter.java (hunk 4 above)
Comment (marqueslanauja) [reply psse/psse-converter/src/main/java/com/powsybl/psse/converter/FixedShuntCompensatorConverter.java]: Done

File: psse/psse-converter/src/main/java/com/powsybl/psse/converter/GeneratorConverter.java (hunk 5 above)
Comment (marqueslanauja) [reply psse/psse-converter/src/main/java/com/powsybl/psse/converter/GeneratorConverter.java]: Done

File: psse/psse-converter/src/main/java/com/powsybl/psse/converter/SwitchedShuntCompensatorConverter.java (hunk 11 above)
Comment (marqueslanauja) [reply psse/psse-converter/src/main/java/com/powsybl/psse/converter/SwitchedShuntCompensatorConverter.java]: Done

File: psse/psse-converter/src/main/java/com/powsybl/psse/converter/TransformerConverter.java (hunk 12 above)
Comment (marqueslanauja) [reply psse/psse-converter/src/main/java/com/powsybl/psse/converter/TransformerConverter.java]: Done

File: psse/psse-converter/src/main/java/com/powsybl/psse/converter/TransformerConverter.java (hunk 13 above)
Comment (marqueslanauja) [reply psse/psse-converter/src/main/java/com/powsybl/psse/converter/TransformerConverter.java]: Done

File: psse/psse-converter/src/main/java/com/powsybl/psse/converter/TransformerConverter.java (hunk 14 above)
Comment (marqueslanauja) [reply psse/psse-converter/src/main/java/com/powsybl/psse/converter/TransformerConverter.java]: Deleted. I kept emphasizing that there is not conversion associated to the angle

File: psse/psse-converter/src/main/java/com/powsybl/psse/converter/VoltageLevelConverter.java (hunk 15 above)
Comment (marqueslanauja) [reply psse/psse-converter/src/main/java/com/powsybl/psse/converter/VoltageLevelConverter.java]: Done

File: psse/psse-converter/src/main/java/com/powsybl/psse/converter/LineConverter.java (hunk 6 above)
Comment (marqueslanauja) [reply psse/psse-converter/src/main/java/com/powsybl/psse/converter/LineConverter.java]: Done

File: psse/psse-converter/src/main/java/com/powsybl/psse/converter/LineConverter.java (hunk 7 above)
Comment (marqueslanauja) [reply psse/psse-converter/src/main/java/com/powsybl/psse/converter/LineConverter.java]: Done

File: psse/psse-converter/src/main/java/com/powsybl/psse/converter/LoadConverter.java (hunk 8 above)
Comment (marqueslanauja) [reply psse/psse-converter/src/main/java/com/powsybl/psse/converter/LoadConverter.java]: Done

File: psse/psse-converter/src/main/java/com/powsybl/psse/converter/SlackConverter.java (hunk 9 above)
Comment (marqueslanauja) [reply psse/psse-converter/src/main/java/com/powsybl/psse/converter/SlackConverter.java]: Done

File: psse/psse-converter/src/main/java/com/powsybl/psse/converter/SubstationConverter.java (hunk 10 above)
Comment (marqueslanauja) [reply psse/psse-converter/src/main/java/com/powsybl/psse/converter/SubstationConverter.java]: Done

File: psse/psse-converter/src/main/java/com/powsybl/psse/converter/BusConverter.java [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+
+/**
+ * @author Luma Zamarreño <zamarrenolm at aia.es>
+ * @author José Antonio Marqués <marquesja at aia.es>
+ */
+public class BusConverter extends AbstractConverter {
+
+    public BusConverter(PsseBus psseBus, ContainersMapping containerMapping, Network network) {
+        super(containerMapping, network);
+        this.psseBus = psseBus;
```
[the remark this answers is not in the document]
Comment (marqueslanauja) [reply psse/psse-converter/src/main/java/com/powsybl/psse/converter/BusConverter.java]: Done

**Comment by MioRtia** [comment]
> Just a question about your new classes: is this necessary for them to be public or can they be package private? If they can, you should put them at the most restricted mode

Could you please look into this?

**Pull request merged**